### DISCUSSION TITLE
An SRG is packed into a constant buffer?

### DISCUSSION THREAD

**Discussion by ohmaya** [o3de/o3de#13402] [General]
https://www.o3de.org/docs/atom-guide/glossary/
"A Shader Resource Group (SRG) is a collection of shader resources (textures, buffers, samplers, and loose constants) that are automatically **packed into an implicit constant buffer**."

Textures, buffers, and samplers do not appear in constant buffers, right?
SRGs map to descriptor tables or descriptor set layouts, right?

**Comment by jeremyong-az** [comment]
Yea that statement seems inaccurate or potentially there was a typo. The top-level declarations of data fields within an SRG are packed into an implicit struct which is bound with the SRG as a constant buffer. The textures and buffers are bound in descriptor tables/sets correspond to an inferred binding scheme, and the sampler is embedded in the pipeline state.

**Comment by galibzon** [comment]
Correct, the glossary needs a more accurate description. Even though in AZSL you must declare textures, buffers, samplers, and variables (of fundamental types or structs) inside SRGs, in the end all resources will have their own binding locations. Only the variables will remain in an implicit Constant Buffer that is created on behalf of the SRG.

For example this AZSL snippet:
```
ShaderResourceGroupSemantic slot1
{
    FrequencyId = 1;
};

ShaderResourceGroup MySRG : slot1
{
    float m_value;
    Texture2D m_tex;
};

```
Will compile to (using AZSLc):  
```
Texture2D MySRG_m_tex : register(t0);

struct MySRG_SRGConstantsStruct
{
     float MySRG_m_value;
};

ConstantBuffer<::MySRG_SRGConstantsStruct> MySRG_SRGConstantBuffer : register(b0);
```
